File: lifeos-bot/ai.py
"""
LifeOS AI Provider Abstraction Layer
=====================================
Swap between Groq, Gemini, Ollama, and Claude
with a single environment variable: AI_PROVIDER

Portfolio note: This pattern is called the "Strategy Pattern" —
a classic software design pattern where the algorithm (AI provider)
is swappable at runtime without changing the calling code.
"""

import os
import json
import httpx
from dotenv import load_dotenv

load_dotenv()

AI_PROVIDER = os.getenv("AI_PROVIDER", "groq")  # groq | gemini | ollama | claude
GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", "")
CLAUDE_API_KEY = os.getenv("CLAUDE_API_KEY", "")
OLLAMA_BASE_URL = os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
OLLAMA_MODEL = os.getenv("OLLAMA_MODEL", "llama3.2")


# ─── PROVIDER IMPLEMENTATIONS ────────────────────────────────────────────────

async def _groq_chat(messages: list[dict], system: str) -> str:
    """Groq — free tier, Llama 3.3 70B, fastest inference"""
    async with httpx.AsyncClient(timeout=30) as client:
        response = await client.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers={"Authorization": f"Bearer {GROQ_API_KEY}"},
            json={
                "model": "llama-3.3-70b-versatile",
                "messages": [{"role": "system", "content": system}, *messages],
                "max_tokens": 1024,
                "temperature": 0.7,
            },
        )
        response.raise_for_status()
        return response.json()["choices"][0]["message"]["content"]


async def _gemini_chat(messages: list[dict], system: str) -> str:
    """Gemini Flash — free tier, 1500 req/day, great for long context"""
    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={GEMINI_API_KEY}"

    # Gemini uses a different message format
    contents = []
    if system:
        contents.append({"role": "user", "parts": [{"text": system}]})
        contents.append({"role": "model", "parts": [{"text": "Understood."}]})
    for msg in messages:
        role = "model" if msg["role"] == "assistant" else "user"
        contents.append({"role": role, "parts": [{"text": msg["content"]}]})

    async with httpx.AsyncClient(timeout=30) as client:
        response = await client.post(url, json={"contents": contents})
        response.raise_for_status()
        return response.json()["candidates"][0]["content"]["parts"][0]["text"]


async def _ollama_chat(messages: list[dict], system: str) -> str:
    """Ollama — fully local, runs on your uni GPU, zero cost"""
    async with httpx.AsyncClient(timeout=60) as client:
        response = await client.post(
            f"{OLLAMA_BASE_URL}/api/chat",
            json={
                "model": OLLAMA_MODEL,
                "messages": [{"role": "system", "content": system}, *messages],
                "stream": False,
            },
        )
        response.raise_for_status()
        return response.json()["message"]["content"]


async def _claude_chat(messages: list[dict], system: str) -> str:
    """Claude Haiku — best reasoning, use for complex scheduling logic"""
    async with httpx.AsyncClient(timeout=30) as client:
        response = await client.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key": CLAUDE_API_KEY,
                "anthropic-version": "2023-06-01",
            },
            json={
                "model": "claude-haiku-4-5-20251001",
                "max_tokens": 1024,
                "system": system,
                "messages": messages,
            },
        )
        response.raise_for_status()
        return response.json()["content"][0]["text"]


# ─── PUBLIC INTERFACE ─────────────────────────────────────────────────────────

async def chat(messages: list[dict], system: str = "") -> str:
    """
    Main entry point. Calls whichever provider is set in AI_PROVIDER.

    Usage:
        response = await chat(
            messages=[{"role": "user", "content": "Plan my day"}],
            system="You are LifeOS, a personal productivity assistant."
        )
    """
    providers = {
        "groq": _groq_chat,
        "gemini": _gemini_chat,
        "ollama": _ollama_chat,
        "claude": _claude_chat,
    }

    provider_fn = providers.get(AI_PROVIDER)
    if not provider_fn:
        raise ValueError(f"Unknown AI_PROVIDER: {AI_PROVIDER}. Choose from: {list(providers.keys())}")

    return await provider_fn(messages, system)


async def extract_goal_from_text(user_text: str, existing_projects: list) -> dict:
    """
    Takes a freeform message like:
    "I have CNN coursework due March 20, need to implement the model and write the report"

    Returns structured JSON:
    {
      "title": "CNN Coursework",
      "area_id": "study",
      "deadline": "2025-03-20",
      "priority": "high",
      "urgency": "high",
      "estimated_hours": 15,
      "difficulty": "hard",
      "tasks": [
        {"title": "Implement CNN model", "type": "coding", "estimated_mins": 180, "difficulty": "hard"},
        {"title": "Write report", "type": "writing", "estimated_mins": 240, "difficulty": "medium"}
      ],
      "clarifying_questions": []  // empty if enough info, otherwise ask these
    }
    """
    system = """You are LifeOS, an AI productivity assistant. 
Extract structured goal/project data from the user's message.
Return ONLY valid JSON, no markdown, no explanation.

Life areas to categorise into:
work, study, health, social, hobbies, selfcare, sleep, meals, growth, admin

Task types: coding, writing, studying, meeting, exercise, reading, design, admin, other

Difficulty levels: easy, medium, hard
Priority levels: low, medium, high
Urgency levels: low, medium, high

If information is missing (deadline, hours estimate, difficulty), 
add it to clarifying_questions as a list of strings.
If enough info exists, clarifying_questions should be empty."""

    existing_titles = [p.get("title", "") for p in existing_projects]
    prompt = f"""User message: "{user_text}"

Existing projects (for context, avoid duplicates): {existing_titles}

Extract and return JSON in this exact format:
{{
  "title": "project title",
  "area_id": "study",
  "deadline": "YYYY-MM-DD or null",
  "priority": "high",
  "urgency": "high", 
  "estimated_hours": 15,
  "difficulty": "hard",
  "tasks": [
    {{"title": "task name", "type": "coding", "estimated_mins": 180, "difficulty": "hard", "due_date": "YYYY-MM-DD or null"}}
  ],
  "clarifying_questions": []
}}"""

    response = await chat(
        messages=[{"role": "user", "content": prompt}],
        system=system
    )

    # Strip markdown fences if model adds them
    cleaned = response.strip().removeprefix("```json").removeprefix("```").removesuffix("```").strip()
    return json.loads(cleaned)


async def generate_daily_schedule(projects: list, tasks: list, date: str) -> list:
    """
    Generates a time-blocked daily schedule from active projects/tasks.

    Returns list of time blocks:
    [
      {"time": "09:00", "end": "10:30", "type": "deep", "title": "...", "task_id": "..."},
      {"time": "10:30", "end": "10:45", "type": "break", "title": "Break"},
      ...
    ]
    """
    system = """You are LifeOS scheduler. Generate an optimal time-blocked daily schedule.
Rules:
- Start at 09:00, end by 22:00
- Deep work blocks: 90 mins max, then 15 min break
- Prioritise high urgency + close deadlines first
- Mix difficult and easy tasks (dopamine scheduling)
- Include meal breaks: lunch 12:30-13:30, dinner 18:00-18:30
- Return ONLY valid JSON array, no markdown."""

    # Summarise tasks for the prompt
    task_summary = []
    for t in tasks:
        if not t.get("done"):
            task_summary.append({
                "id": t.get("id"),
                "title": t.get("title"),
                "type": t.get("task_type", "general"),
                "estimated_mins": t.get("estimated_mins", 30),
                "difficulty": t.get("difficulty", "medium"),
                "due_date": t.get("due_date"),
                "project": next((p["title"] for p in projects if p["id"] == t.get("project_id")), "")
            })

    prompt = f"""Date: {date}
Pending tasks: {json.dumps(task_summary[:10])}  

Generate time-blocked schedule as JSON array:
[
  {{"time": "09:00", "end": "10:30", "type": "deep", "title": "task title", "task_id": "uuid or null", "area": "study"}},
  {{"time": "10:30", "end": "10:45", "type": "break", "title": "Break", "task_id": null, "area": null}}
]

Block types: deep, medium, quick, break, meal, reflection"""

    response = await chat(
        messages=[{"role": "user", "content": prompt}],
        system=system
    )

    cleaned = response.strip().removeprefix("```json").removeprefix("```").removesuffix("```").strip()
    return json.loads(cleaned)


async def generate_motivational_checkin(task_title: str, response_type: str, context: dict) -> str:
    """
    Generate a supportive check-in response based on user's answer.
    response_type: done | almost | not_started | distracted
    """
    system = """You are LifeOS, a supportive productivity coach. 
Keep responses SHORT (2-3 sentences max), warm, and action-oriented.
Never be preachy. Be like a smart friend who gets it."""

    prompts = {
        "done": f"User completed: '{task_title}'. Celebrate briefly and mention XP earned. Streak: {context.get('streak', 0)} days.",
        "almost": f"User almost finished: '{task_title}'. Encourage, suggest 30 min catch-up block.",
        "not_started": f"User didn't start: '{task_title}'. Be understanding, help reschedule without guilt.",
        "distracted": f"User got distracted during: '{task_title}'. Empathise, suggest a fresh start strategy.",
    }

    response = await chat(
        messages=[{"role": "user", "content": prompts.get(response_type, prompts["done"])}],
        system=system
    )
    return response

File: lifeos-web/src/lib/ai.js
/**
 * LifeOS — AI Provider Layer (Frontend)
 * =======================================
 * Mirrors the Python ai.py but runs in the browser.
 * Provider is set by the user in Settings and stored in localStorage.
 *
 * Portfolio note: Same Strategy Pattern as the Python layer.
 * Consistent architecture across both surfaces.
 */

// ─── CONFIG ───────────────────────────────────────────────────────────────────

export const AI_PROVIDERS = {
  groq: {
    name: "Groq · Llama 3.3",
    label: "Free",
    color: "#f59e0b",
    model: "llama-3.3-70b-versatile",
    baseURL: "https://api.groq.com/openai/v1/chat/completions",
  },
  gemini: {
    name: "Gemini 1.5 Flash",
    label: "Free",
    color: "#3b82f6",
    model: "gemini-1.5-flash",
    baseURL: "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent",
  },
  claude: {
    name: "Claude Haiku",
    label: "~£0.01/msg",
    color: "#a78bfa",
    model: "claude-haiku-4-5-20251001",
    baseURL: "https://api.anthropic.com/v1/messages",
  },
  ollama: {
    name: "Ollama (Local)",
    label: "Free",
    color: "#34d399",
    baseURL: "http://localhost:11434/api/chat",
  },
};

// ─── CONFIG STORAGE ───────────────────────────────────────────────────────────

export function getAIConfig() {
  return {
    provider: localStorage.getItem("ai_provider") || "groq",
    apiKey: localStorage.getItem("ai_api_key") || "",
    ollamaModel: localStorage.getItem("ollama_model") || "llama3.2",
  };
}

export function saveAIConfig(config) {
  localStorage.setItem("ai_provider", config.provider);
  localStorage.setItem("ai_api_key", config.apiKey);
  if (config.ollamaModel) localStorage.setItem("ollama_model", config.ollamaModel);
}

// ─── PROVIDER IMPLEMENTATIONS ─────────────────────────────────────────────────

async function groqChat(messages, system, apiKey) {
  const response = await fetch("https://api.groq.com/openai/v1/chat/completions", {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}`, "Content-Type": "application/json" },
    body: JSON.stringify({
      model: "llama-3.3-70b-versatile",
      messages: [{ role: "system", content: system }, ...messages],
      max_tokens: 1024,
      temperature: 0.7,
    }),
  });
  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.error?.message || `Groq error ${response.status}`);
  }
  const data = await response.json();
  return data.choices[0].message.content;
}

async function geminiChat(messages, system, apiKey) {
  const contents = [];
  if (system) {
    contents.push({ role: "user", parts: [{ text: system }] });
    contents.push({ role: "model", parts: [{ text: "Understood." }] });
  }
  for (const msg of messages) {
    contents.push({
      role: msg.role === "assistant" ? "model" : "user",
      parts: [{ text: msg.content }],
    });
  }
  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${apiKey}`,
    { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ contents }) }
  );
  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.error?.message || `Gemini error ${response.status}`);
  }
  const data = await response.json();
  return data.candidates[0].content.parts[0].text;
}

async function claudeChat(messages, system, apiKey) {
  const response = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
      "anthropic-dangerous-direct-browser-access": "true",
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: "claude-haiku-4-5-20251001",
      max_tokens: 1024,
      system,
      messages,
    }),
  });
  if (!response.ok) {
    const err = await response.json();
    throw new Error(err.error?.message || `Claude error ${response.status}`);
  }
  const data = await response.json();
  return data.content[0].text;
}

async function ollamaChat(messages, system, ollamaModel = "llama3.2") {
  const response = await fetch("http://localhost:11434/api/chat", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model: ollamaModel,
      messages: [{ role: "system", content: system }, ...messages],
      stream: false,
    }),
  });
  if (!response.ok) throw new Error(`Ollama error ${response.status} — is Ollama running?`);
  const data = await response.json();
  return data.message.content;
}

// ─── PUBLIC INTERFACE ─────────────────────────────────────────────────────────

export async function chat(messages, system = "") {
  const { provider, apiKey, ollamaModel } = getAIConfig();

  switch (provider) {
    case "groq":    return groqChat(messages, system, apiKey);
    case "gemini":  return geminiChat(messages, system, apiKey);
    case "claude":  return claudeChat(messages, system, apiKey);
    case "ollama":  return ollamaChat(messages, system, ollamaModel);
    default:        throw new Error(`Unknown provider: ${provider}`);
  }
}

export async function extractGoalFromText(userText, existingProjects = []) {
  const today = new Date().toISOString().split("T")[0];
  const system = `You are LifeOS, an AI productivity assistant.
Today's date is ${today}. Always use this as reference for relative dates like "next Friday" or "tomorrow".
Extract structured goal/project data from the user's message.
Return ONLY valid JSON, no markdown, no explanation.

Life area IDs: work, study, health, social, hobbies, selfcare, sleep, meals, growth, admin
Task types: coding, writing, studying, meeting, exercise, reading, design, admin, other
Difficulty: easy | medium | hard
Priority: low | medium | high`;

  const existingTitles = existingProjects.map((p) => p.title);
  const prompt = `User message: "${userText}"
Existing projects (avoid duplicates): ${JSON.stringify(existingTitles)}

Return this exact JSON structure:
{
  "title": "project title",
  "area_id": "study",
  "deadline": "YYYY-MM-DD or null",
  "priority": "high",
  "urgency": "high",
  "estimated_hours": 15,
  "difficulty": "hard",
  "tasks": [
    {"title": "task name", "type": "coding", "estimated_mins": 180, "difficulty": "hard", "due_date": "YYYY-MM-DD or null"}
  ],
  "clarifying_questions": []
}`;

  const response = await chat([{ role: "user", content: prompt }], system);
  const cleaned = response.trim().replace(/^```json\n?/, "").replace(/^```\n?/, "").replace(/```$/, "").trim();
  return JSON.parse(cleaned);
}

export async function generateDailySchedule(projects, tasks) {
  const today = new Date().toISOString().split("T")[0];
  const system = `You are LifeOS scheduler. Generate an optimal time-blocked daily schedule.
Rules:
- Start 09:00, end by 22:00. Deep work max 90 mins, then 15 min break.
- Prioritise high urgency + nearest deadlines first.
- Mix difficult and easy tasks (dopamine scheduling).
- Include lunch 12:30-13:30.
- Return ONLY a valid JSON array, no markdown.`;

  const taskSummary = tasks
    .filter((t) => !t.done)
    .slice(0, 10)
    .map((t) => ({
      id: t.id,
      title: t.title,
      type: t.task_type || "general",
      estimated_mins: t.estimated_mins || 30,
      difficulty: t.difficulty || "medium",
      due_date: t.due_date,
      project: projects.find((p) => p.id === t.project_id)?.title || "",
    }));

  const prompt = `Date: ${today}
Pending tasks: ${JSON.stringify(taskSummary)}

Return schedule as JSON array:
[{"time":"09:00","end":"10:30","type":"deep","title":"...","task_id":"uuid or null","area":"study"}]
Block types: deep | medium | quick | break | meal | reflection`;

  const response = await chat([{ role: "user", content: prompt }], system);
  const cleaned = response.trim().replace(/^```json\n?/, "").replace(/^```\n?/, "").replace(/```$/, "").trim();
  return JSON.parse(cleaned);
}
